redhat-appstudio/remote-secret
# Add the ability use a different identity than the operator SA when deploying to targets

### What does this PR do?
This improves the security aspects of the code because we can now limit where the remote secrets can be deployed to on per-namespace basis.

Because of the tests for the new functionality in controllers/client_factory_test.go, this also updates the controller-runtime to v0.15.0 which caused some other collateral changes in otherwise unrelated parts of the codebase.

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-546

### How to test this PR?

#### Auth using kubeconfig

1. Have a remote OpenShift cluster ready and a local Minikube instance.
1. In OpenShift, have an admin user and some unprivileged user.
1. Log in to the OpenShift cluster using the admin account and create a namespace `test-target-namespace`.
1. Create a role that can crud secrets and service accounts.
   ```yaml
   kind: Role       
   apiVersion: rbac.authorization.k8s.io/v1
   metadata:
     name: can-crud-secrets      
     namespace: test-target-namespace
   rules:  
   - apiGroups:                         
     - ""      
     resources:            
     - "secrets"
     - "serviceaccounts"
     verbs:    
     - "list"   
     - "get"
     - "create"
     - "update"
     - "delete"
   ```
1. Create a role binding to that role for the unprivileged user:
   ```yaml
   kind: RoleBinding
   apiVersion: rbac.authorization.k8s.io/v1
   metadata:
     name: user1-can-crud-secrets
     namespace: test-target-namespace
   roleRef:
     apiGroup: rbac.authorization.k8s.io
     kind: Role
     name: can-crud-secrets
   subjects:
   - kind: User
     name: <<< NAME OF THE UNPRIVILEGED USER >>>
   ```
1. Login to the OpenShift cluster from the command line using the unprivileged user. Take note of the name of the context and server url for this cluster and user from your `$KUBECONFIG`.
2. Start minikube. 
3. Create a secret with a kubeconfig for the remote OpenShift cluster:
   `hack/create-secret-with-kubeconfig.sh '<<OPENSHIFT_CONTEXT>>' test-remote-kubeconfig default`
4. Create a remote secret (make sure to edit the YAML file with the actual URL of your openshift cluster:
   `kubectl apply -f samples/remote-cluster-secret.yaml`
1. Upload the data to the remote secret:
   `kubectl apply -f samples/remote-cluster-secret-secret.yaml`
1. Observe that the secrets and the service accounts were created in the OpenShift cluster in the `test-target-namespace`.
1. Change the remote secret so that the target points to the `default` namespace in the OpenShift cluster.
1. Notice that the secret and service account disappeared from the the `test-target-namespace` but the remote secret now reports an error that it could deploy because of insufficient permissions.

#### Auth using service account in the local cluster
1. Have just Minikube
2. Create `test-target-namespace`
1. Create the following service account in the default namespace:
	```yaml
	apiVersion: v1
	kind: ServiceAccount
	metadata:
          name: deployment-enabler
	  namespace: default
	  labels:
	    appstudio.redhat.com/remotesecret-auth-sa: "true"
	```
1. Create a cluster role for enabling secret and serviceaccount crud (this can also be just a role in the `test-target-namespace`):
	```yaml
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRole
	metadata:
	  name: sec-sa-crud
	rules:
	- apiGroups: 
	  - ""
	  resources:
	  - "secrets"
	  - "serviceaccounts"
	  verbs:
	  - "create"
	  - "get"
	  - "list"
	  - "update"
	  - "delete"
	```
1. Bind that cluster role to the `deployment-enabler` SA in the `test-target-namespace` namespace:
	```yaml
	apiVersion: rbac.authorization.k8s.io/v1
	kind: RoleBinding
	metadata:
	  name: "deployment-enabler"
	  namespace: "test-target-namespace"
	roleRef:
	  apiGroup: rbac.authorization.k8s.io   
	  kind: ClusterRole
	  name: sec-sa-crud
	subjects:
	- kind: ServiceAccount
	  name: "deployment-enabler"
	  namespace: "default"	
	```
1. Create a test binding:
   `kubectl apply -f samples/remote-secret.yaml`
1. Upload the data to it:
   `kubectl apply -f samples/remote-secret-secret.yaml`
1. The secret and service account should appear in the `test-target-namespace`.
1. Create a new namespace `another-target-namespace`.
1. Add a new target in the remote secret that points to the `another-target-namespace`.
1. Notice that the deployment to `another-target-namespace` failed because of the lack of permissions.

#### Target in the current namespace
1. Have a Minikube
1. Create a remote secret with a target pointing to the same namespace as the remote secret lives in.
1. Upload the secret data for the remote secret.
1. Notice that the deployment succeeded.
